File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/scala/lists/P09.scala
package com.shekhargulati.ninetynine_problems.scala.lists

/**
  * <b>(**) Pack consecutive duplicates of list elements into sublists</b>
  * <p>If a list contains repeated elements they should be placed in separate sublists.</p>
  *
  * <pre>
  * pack(Arrays.asList("a", "a", "a", "a", "b", "c", "c", "a", "a", "d", "e", "e", "e", "e"))
  * [["a","a","a","a"],["b"],["c","c"],["a","a"],["d"],["e","e","e","e"]]
  * </pre>
  */
object P09 {


  def pack[T](list: List[T]): List[List[T]] = {
    list match {
      case x :: xs => (x +: xs.takeWhile(el => el == x)) +: pack(xs.dropWhile(el => el == x))
      case Nil => Nil
    }
  }

}

File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P09.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * <b>(**) Pack consecutive duplicates of list elements into sublists</b>
 * <p>If a list contains repeated elements they should be placed in separate sublists.</p>
 *
 * <pre>
 *     pack(Arrays.asList("a", "a", "a", "a", "b", "c", "c", "a", "a", "d", "e", "e", "e", "e"))
 *     [["a","a","a","a"],["b"],["c","c"],["a","a"],["d"],["e","e","e","e"]]
 * </pre>
 */
public class P09 {

    /*
    Simple approach using for-each loop
     */
    public static <T> List<List<T>> pack(List<T> list) {
        T lastElement = null;
        List<List<T>> packedList = new ArrayList<>();
        List<T> elements = new ArrayList<>();
        for (T el : list) {
            if (!Objects.equals(lastElement, el)) {
                elements = new ArrayList<>();
                packedList.add(elements);
            }
            elements.add(el);
            lastElement = el;
        }
        return packedList;
    }


}
